### ISSUE TITLE
Duplicated mounted call on component in Vue story

### ISSUE THREAD

**Issue by mattiaerli97** [storybookjs/storybook#13633]
**Describe the bug**
When I include a component in a story, the 'mounted' lifecycle is called twice. This doesn't happen when the component is used outside Storybook

**Expected behavior**
I expect that the 'mounted' lifecycle is called only once

**Code snippets**
This is my code. I'm following the structure tu use 'CONTROLS'

```import StarRating from './index.vue';
import Docs from './index.mdx';
import { action } from '@storybook/addon-actions';

export default { title: '2. Molecules/StarRating', component: StarRating };

// Parameters of the story
const parametersStory = {
  controls: { expanded: true, hideNoControlsWarning: true },
  docs: { page: Docs },
  design: { disable: true },
  jest: ['starRating.test.js']
};

// Argtypes for controls
const argTypesStory = {
  skillLevel: { control: { disable: true } },
  levels: { control: { disable: true } }
};

// Actions on events
const actionsData = {
  onLevelSelected: action('onLevelSelected')
};

// The template with args and argTypes
const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { StarRating },
  template:
    '<StarRating\n' +
    ' :disabled="disabled"' +
    ' :visibleEmptyStar="visibleEmptyStar"' +
    ' :levels="levels"\n' +
    ' :skillLevel="skillLevel"\n' +
    ' @onLevelSelected="onLevelSelected"\n' +
    '></StarRating>'
});

// The Default component that extends the Template
export const Default = Template.bind({});
Default.args = Object.assign(
  {},
  {
    levels: [1, 2, 3, 4, 5],
    skillLevel: 2
  },
  actionsData
);

// Final declarations to use the story
Default.parameters = parametersStory;
Default.argTypes = argTypesStory;
```

In the 'Template' declaration, if I didn't include the part `args, { argTypes }` it mounts only once, as expected. However, without that code, controls don't work.

**Version**
I'm using the Storybook version 6.1.14 and the addons have all the same version

Could you please help me to find a solution to this struggling problem? Thanks a lot

**Comment by pocka** [comment]
I think this is due to Dynamic Source Rendering mounts the component in order to obtain runtime props. ([source code](https://github.com/storybookjs/storybook/blob/afd0eb57db4355467a06c353f028cbf99c2f4095/addons/docs/src/frameworks/vue/sourceDecorator.ts#L34-L48))

The easiest workaround is to disable the Dynamic Source Rendering by setting `type: "code"` to `docs.source`.

```diff
- export default { title: '2. Molecules/StarRating', component: StarRating };
+ export default {
+   title: '2. Molecules/StarRating',
+   component: StarRating,
+   parameters: { docs: { source: { type: "code" }}}
+ };
```

Unless you put a side-effect that affects outside the Vue tree things, you can ignore that (because it's happening in a separated tree).

**Comment by mattiaerli97** [comment]
Hi @pocka!
Thank you for the solution, it worked for me!
Could you please explain me the reasoning behind the solution? I'm curious. Is it only happening in Vue?

**Comment by pocka** [comment]
@mattiaerli97 
Yes, this is specific to Vue.js with Dynamic Source Rendering. The Docs Addon switches to Dynamic Source Rendering when the story is Args Story (stories takes one or more arguments). The Dynamic Source Rendering feature tries to build "live" code snippets by serializing VNode (Vue's virtual DOM tree data/node).

The `{ type: "code" }` parameter disables the Dynamic Source Rendering feature and it renders a story source code.

The reason we mount a story for this is, the only way to grab the real-time VNode is to mount the story in the decorator AFAIK.
I know this comes with downsides and limitations but this is the most straightforward and low-effort way to archive the goal. I think it's *just okay* if we provide a workaround or escape hatch. I hope someone will implement the feature for Vue3 in a better way... :tanabata_tree:

**Comment by mattiaerli97** [comment]
Thanks for the explanation @pocka, I think we can close the issue

**Comment by mattiaerli97** [comment]
Issue closed, as in Vue there isn't any other solution

**Comment by kaidjohnson** [comment]
Not sure if it's worth opening a new ticket, but this issue affects components that render elements outside of their tree (modals, for example), by rendering the component twice. As a result, any Vue Storybook that renders elements such as modals would benefit from the workaround outlined above, which can also be fixed globally from `.storybook/preview.js`:

```
export const parameters = {
  docs: { 
    source: { 
      type: "code" 
    }
  }
};
```

I set up a repo to demonstrate the problem: https://github.com/kaidjohnson/storybook-issue-13633

Note: This issue was not visible in Storybook@6.0.28, so it feels like a regression in Storybook@6.1.x. I'm not sure waiting for Vue3 is a good solution, as many teams will remain on Vue2 for some time to come and this issue might prevent them from upgrading Storybook past 6.0.28. At the very least, should this workaround be included as part of the core storybook-vue package or in its documentation somewhere?

**Comment by shilman** [comment]
¡Ay Caramba!! I just released https://github.com/storybookjs/storybook/releases/tag/v6.2.0-beta.2 containing PR #14002 that references this issue. Upgrade today to the `@next` NPM tag to try it out!

```
npx sb upgrade --prerelease
```

**Comment by clorichel** [comment]
Vue 3 users here 🤓  And our components library is one of many packages in a monorepo. Storybook working fine on first load ✌️  **but** we are unable to change story without getting an empty canvas and this in our browser console:

```
[Vue warn]: App has already been mounted.
If you want to remount the same app, move your app creation logic into a factory function and create fresh app instances for each mount - e.g. `const createMyApp = () => createApp(App)`
```

We tried:

1. the `docs: { source: { type: "code" }}` trick in `.storybook/preview.js` with no luck. Even got to trying `docs: { page: null }` with no success either 🙃 
2. upgraded to the prerelease `"@storybook/vue3": "^6.4.0-alpha.33"` but we still cannot browse our stories, while facing the exact same issue: first story to be loaded displays fine, clicking on another story shows the error in the console and a white canvas, refreshing the page with the browser does make it appear though (inconveniently).

Weirdest thing: if we duplicate the component library package **outside of our monorepo**, remove node_modules and `yarn install` it before `yarn storybook`: it does work just fine! 😲  Not quite sure from here if we should open another issue, hopefully someone here would have insights ❤️

**Comment by shilman** [comment]
@clorichel please open a new issue! 🙏
